### PR TITLE
Session monitoring enhancement

### PR DESCRIPTION
**Pull request by burgerkingeater**
1. The metrics being added shows the number of user sessions in the system, which can be leveraged to determine the session cache size appropriately.
2. Logging showing the number of sessions with same IP is added to help detect DDoS.

### CODE REVIEW

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java [hunk 1]
```diff
@@ -409,6 +410,10 @@ protected void handleAjaxLoginAction(final HttpServletRequest req,
       cookie.setPath("/");
       resp.addCookie(cookie);
       getApplication().getSessionCache().addSession(session);
+      final Set<Session> sessionsOfSameIP =
+          getApplication().getSessionCache().findSessionsByIP(session.getIp());
+      // Check potential DDoS attack by bad hosts.
+      logger.info(sessionsOfSameIP.size() + " session(s) found from this IP " + session.getIp());
```
Comment (ypadron-in) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java]: Can we add the log message when there are more than one sessions per ip?
BTW, not all sessions are registered in SessionCache. Here `LoginAbstractAzkabanServlet.doPost()` sessions are temporarily created but are not registered. Do you know the reason behind it?

File: azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java [hunk 2]
```diff
@@ -71,6 +73,10 @@ private void setupAllMetrics() {
     this.queueWaitMeter = this.metricsManager.addHistogram(QUEUE_WAIT_HISTOGRAM_NAME);
   }
 
+  public <T> void addSessionCount(final Supplier<T> getSessionCount) {
+    this.metricsManager.addGauge(SESSION_COUNT, getSessionCount);
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/metrics/CommonMetrics.java]: Have you thought about registering this metric after SessionCache's initialization, outside its constructor? Maybe SessionCache doesn't need to know about metrics existence.

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java [hunk 3]
```diff
@@ -46,13 +52,16 @@
...
+  public SessionCache(final Props props, @Nullable final CommonMetrics metrics) {
     this.effectiveSessionTimeToLive = props.getLong(ConfigurationKeys.SESSION_TIME_TO_LIVE,
         DEFAULT_SESSION_TIME_TO_LIVE);
     this.cache = CacheBuilder.newBuilder()
         .maximumSize(props.getInt("max.num.sessions", MAX_NUM_SESSIONS))
-        .expireAfterAccess(effectiveSessionTimeToLive, TimeUnit.MILLISECONDS)
+        .expireAfterAccess(this.effectiveSessionTimeToLive, TimeUnit.MILLISECONDS)
         .build();
+    if (metrics != null) {
+      metrics.addSessionCount(this::getSessionCount);
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: Have you thought about registering this metric after SessionCache's initialization, outside its constructor? Maybe SessionCache doesn't need to know about metrics existence.

Comment (burgerkingeater) [reply azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: good point, updated.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java (hunk 1 above)
Comment (burgerkingeater) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java]: enrich the logging to have more info.

Temp session is used when user specify username/password in their non-login request. But I do think this is an anti-pattern in terms of security. Let's see if we can deprecate this behavior.

File: azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java [hunk 4]
```diff
@@ -107,7 +108,7 @@ public static MockLoginAzkabanServlet getServletWithSession(final String session
 
     // Create a concrete SessionCache so a session will get persisted
     // and can get looked up
-    final SessionCache cache = new SessionCache(props);
+    final SessionCache cache = new SessionCache(props, null);
```
Comment (ypadron-in) [inline azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java]: Can you update constructor parameters here?

Comment (burgerkingeater) [reply azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java]: thanks, updated

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java [hunk 5]
```diff
@@ -181,6 +181,8 @@ public AzkabanWebServer(final Props props,
     this.scheduler = requireNonNull(scheduler, "scheduler is null.");
     this.flowTriggerService = requireNonNull(flowTriggerService, "flow trigger service is null");
 
+    this.metricsManager.addGauge("session-count", this.sessionCache::getSessionCount);
```
Comment (ypadron-in) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: Should we move this line to startWebMetrics()?

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java [hunk 6]
```diff
@@ -409,6 +412,14 @@ protected void handleAjaxLoginAction(final HttpServletRequest req,
...
       resp.addCookie(cookie);
       getApplication().getSessionCache().addSession(session);
+      final Set<Session> sessionsOfSameIP =
+          getApplication().getSessionCache().findSessionsByIP(session.getIp());
+
+      // Check potential DDoS attack by bad hosts.
+      logger.info(
+          "Session id created for user '" + session.getUser().getUserId() + "' and ip " + session
+              .getIp() + ", " + sessionsOfSameIP.size() + " session(s) found from this IP "
+              + session.getIp());
```
Comment (ypadron-in) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java]: Do we need to repeat the ip information?

Comment (burgerkingeater) [reply azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java]: good catch

**Pull request merged**